test/gtest/ucp/test_ucp_tag_perf.cc

```diff
@@ -0,0 +1,81 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
+* Copyright (C) UT-Battelle, LLC. 2015. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "test_ucp_tag.h"
+
+#include <common/test_helpers.h>
+
+
+class test_ucp_tag_perf : public test_ucp_tag {
+public:
+    virtual void init() {
+        if (RUNNING_ON_VALGRIND) {
+            UCS_TEST_SKIP_R("valgrind");
+        }
+        test_ucp_tag::init();
+    }
+
+protected:
+    static const size_t    COUNT    = 10000;
+    static const ucp_tag_t TAG_MASK = 0xffffffffffffffffUL;
+
+    void check_perf(ucs_time_t start_time, double exp_time_ns);
+    void do_sends();
+};
+
+void test_ucp_tag_perf::check_perf(ucs_time_t start_time, double exp_time_ns)
+{
+    ucs_time_t elapsed = ucs_get_time() - start_time;
+    double nsec_per_req = (ucs_time_to_nsec(elapsed) / COUNT);
+    UCS_TEST_MESSAGE << nsec_per_req << " ns per request";
+    EXPECT_LT(nsec_per_req, exp_time_ns) << "Tag matching is not scalable";
+}
+
+void test_ucp_tag_perf::do_sends()
+{
+    for (int i = (int)COUNT - 1; i >= 0; --i) {
+        send_b(NULL, 0, DATATYPE, i);
+    }
+}
+
+UCS_TEST_P(test_ucp_tag_perf, multi_exp) {
+    std::vector<request*> rreqs;
+
+    for (size_t i = 0; i < COUNT; ++i) {
+        request *rreq = recv_nb(NULL, 0, DATATYPE, i, TAG_MASK);
+        ASSERT_TRUE(!UCS_PTR_IS_ERR(rreq));
+        EXPECT_FALSE(rreq->completed);
+        rreqs.push_back(rreq);
+    }
+
+    ucs_time_t start_time = ucs_get_time();
+    do_sends();
+
+    for (size_t i = 0; i < COUNT; ++i) {
+        request *rreq = rreqs.back();
+        rreqs.pop_back();
+        wait_and_validate(rreq);
+    }
+    check_perf(start_time, 1e5);
```

Comment:
not sure it would be enough, for example some CPUs may be slower than others even if running with same clock speed :wink:, i think the current threshold is high enough for most modern systems and still fail if the search is O(n)